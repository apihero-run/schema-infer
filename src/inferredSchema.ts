import { JSONStringFormat } from "@jsonhero/json-infer-types";

type NumberRange = {
  min: number;
  max: number;
};

type InferredUnknown = {
  type: "unknown";
};

type InferredAny = {
  type: "any";
  schemas: Set<InferredSchema>;
};

type InferredBoolean = {
  type: "boolean";
};

type InferredInt = {
  type: "int";
  range: NumberRange;
};

type InferredFloat = {
  type: "float";
  range: NumberRange;
};

type InferredString = {
  type: "string";
  format?: JSONStringFormat;
};

type InferredArray = {
  type: "array";
  items: InferredSchema;
};

type InferredObject = {
  type: "object";
  properties: {
    required: Record<string, InferredSchema>;
    optional: Record<string, InferredSchema>;
  };
};

type InferredNullable = {
  type: "nullable";
  schema: InferredSchema;
};

export type InferredSchema =
  | InferredUnknown
  | InferredAny
  | InferredBoolean
  | InferredInt
  | InferredFloat
  | InferredString
  | InferredArray
  | InferredObject
  | InferredNullable;

export function inferRange(
  value: number,
  range: NumberRange = { min: Number.MAX_SAFE_INTEGER, max: Number.MIN_SAFE_INTEGER },
): NumberRange {
  return {
    min: Math.min(range.min, value),
    max: Math.max(range.max, value),
  };
}
