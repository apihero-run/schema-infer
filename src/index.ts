import { InferredSchema } from "./inferredSchema";
import SchemaInferrer from "./SchemaInference";

export function inferSchema(value: unknown, inference?: SchemaInferrer): SchemaInferrer {
  const schemaInferrer = new SchemaInferrer();
  schemaInferrer.infer(value, inference);
  return schemaInferrer;
}

export function restoreSnapshot(snapshot: InferredSchema): SchemaInferrer {
  return new SchemaInferrer(snapshot);
}

export type { SchemaInferrer };
