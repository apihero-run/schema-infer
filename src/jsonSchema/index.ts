import { match } from "ts-pattern";
import { s, Schema, SchemaBuilder, StringFormat } from "@jsonhero/json-schema-fns";
import { InferredSchema } from "../inferredSchema";
import { JSONStringFormat } from "@jsonhero/json-infer-types";

export function toJSONSchema(inferredSchema: InferredSchema): SchemaBuilder<Schema> {
  return match<InferredSchema, SchemaBuilder<Schema>>(inferredSchema)
    .with({ type: "unknown" }, () => s.$false()) // This should never be reached
    .with({ type: "boolean" }, () => s.boolean())
    .with({ type: "nullable" }, ({ schema }) =>
      schema.type == "unknown"
        ? s.nil()
        : schema.type === "nullable"
        ? toJSONSchema(schema)
        : s.nullable(toJSONSchema(schema)),
    )
    .with({ type: "int" }, () => {
      return s.integer();
    })
    .with({ type: "float" }, () => {
      return s.number();
    })
    .with({ type: "string" }, ({ format }) => {
      const formatString = toJSONStringFormat(format);

      return s.string(formatString ? { format: formatString } : {});
    })
    .with({ type: "array" }, (inferredArray) => {
      const items = toJSONSchema(inferredArray.items);

      return s.array({ items });
    })
    .with({ type: "object" }, (inferredObject) => {
      const requiredProperties = Object.entries(inferredObject.properties.required).map(
        ([key, value]) => s.requiredProperty(key, toJSONSchema(value)),
      );

      const optionalProperties = Object.entries(inferredObject.properties.optional).map(
        ([key, value]) => s.property(key, toJSONSchema(value)),
      );

      return s.object({ properties: requiredProperties.concat(optionalProperties) });
    })
    .with({ type: "any" }, ({ schemas }) => {
      return s.anyOf(...Array.from(schemas).map(toJSONSchema));
    })
    .exhaustive();
}

function toJSONStringFormat(format?: JSONStringFormat): StringFormat | undefined {
  if (!format) {
    return undefined;
  }

  switch (format.name) {
    case "hostname":
      return "hostname";
    case "ip":
      return format.variant == "v4" ? "ipv4" : "ipv6";
    case "uri":
      return "uri";
    case "email":
      return "email";
    case "datetime":
      switch (format.parts) {
        case "datetime":
          return "date-time";
        case "date":
          return "date";
        case "time":
          return "time";
        default:
          return undefined;
      }
    case "uuid":
      return "uuid";
  }

  return undefined;
}
