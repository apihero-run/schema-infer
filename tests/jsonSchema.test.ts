import { Schema } from "@jsonhero/json-schema-fns";
import { readFileSync } from "fs";
import { inferSchema, restoreSnapshot, SchemaInferrer } from "../src";

const toSchema = (s: SchemaInferrer): Schema => s.toJSONSchema({ includeSchema: false });
const readFixture = (s: string): unknown =>
  JSON.parse(readFileSync(`./tests/json/${s}.json`, "utf8").toString());
const fixtureToSchema = (s: string): Schema =>
  inferSchema(readFixture(s)).toJSONSchema({
    includeSchema: false,
  });

describe("toJSONSchema()", () => {
  it("should work with the given options", () => {
    expect(inferSchema(1).toJSONSchema({ includeSchema: true })).toEqual({
      $schema: "https://json-schema.org/draft/2020-12/schema",
      type: "integer",
    });

    expect(inferSchema(1).toJSONSchema({ includeSchema: false })).toEqual({
      type: "integer",
    });
  });
});

describe("toSnapshot()/restoreSnapshot()", () => {
  it("Should allow resuming an inferring session", () => {
    let schema = inferSchema({
      foo: "bar",
    });

    const snapshot = schema.toSnapshot();

    schema = inferSchema(
      {
        bar: "baz",
      },
      restoreSnapshot(snapshot),
    );

    expect(toSchema(schema)).toStrictEqual({
      type: "object",
      properties: {
        foo: { type: "string" },
        bar: { type: "string" },
      },
    });
  });
});

describe("simple types", () => {
  it("should infer nulls", () => {
    expect(toSchema(inferSchema(null))).toEqual({ type: "null" });
  });

  it("should infer boolean", () => {
    const inference = inferSchema(true);

    expect(toSchema(inference)).toStrictEqual({
      type: "boolean",
    });
  });

  it("should infer boolean over multiple iterations", () => {
    let inference = inferSchema(true);
    inference = inferSchema(false, inference);

    expect(toSchema(inference)).toStrictEqual({
      type: "boolean",
    });

    inference = inferSchema(1, inference);

    expect(toSchema(inference)).toStrictEqual({
      anyOf: [{ type: "boolean" }, { type: "integer" }],
    });
  });

  it("should infer int", () => {
    expect(toSchema(inferSchema(1))).toStrictEqual({
      type: "integer",
    });
  });

  it("should infer int over multiple iterations", () => {
    let inference = inferSchema(1);
    inference = inferSchema(2, inference);
    inference = inferSchema(10, inference);

    expect(toSchema(inference)).toStrictEqual({
      type: "integer",
    });
  });

  it("should infer floats as numbers", () => {
    expect(toSchema(inferSchema(1.1))).toStrictEqual({
      type: "number",
    });
  });

  it("should infer floats as numbers even over multiple iterations", () => {
    let inference = inferSchema(1.1);
    inference = inferSchema(2.5, inference);
    inference = inferSchema(10.8, inference);

    expect(toSchema(inference)).toStrictEqual({
      type: "number",
    });

    inference = inferSchema("hello world", inference);

    expect(toSchema(inference)).toStrictEqual({
      anyOf: [{ type: "number" }, { type: "string" }],
    });
  });

  it("should infer floats and ints as numbers", () => {
    let inference = inferSchema(1.1);
    inference = inferSchema(2, inference);

    expect(toSchema(inference)).toStrictEqual({
      type: "number",
    });

    let inferenceIntFirst = inferSchema(1);
    inferenceIntFirst = inferSchema(2.5, inferenceIntFirst);

    expect(toSchema(inferenceIntFirst)).toStrictEqual({
      type: "number",
    });
  });

  it("should infer strings", () => {
    expect(toSchema(inferSchema("hello world"))).toStrictEqual({
      type: "string",
    });

    expect(toSchema(inferSchema("https://google.com/#"))).toStrictEqual({
      type: "string",
      format: "uri",
    });

    expect(toSchema(inferSchema("eric@stackhero.dev"))).toStrictEqual({
      type: "string",
      format: "email",
    });

    expect(toSchema(inferSchema("AEA9CF21-965A-46C0-A4DD-3652B0BDC56D"))).toStrictEqual({
      type: "string",
      format: "uuid",
    });

    expect(toSchema(inferSchema("google.com"))).toStrictEqual({
      type: "string",
      format: "hostname",
    });

    expect(toSchema(inferSchema("192.168.1.0"))).toStrictEqual({
      type: "string",
      format: "ipv4",
    });

    expect(toSchema(inferSchema("2001:db8:1234::1"))).toStrictEqual({
      type: "string",
      format: "ipv6",
    });

    expect(toSchema(inferSchema("2019-01-01 00:00:00.000Z"))).toStrictEqual({
      type: "string",
      format: "date-time",
    });

    expect(toSchema(inferSchema("2016-05-25"))).toStrictEqual({
      type: "string",
      format: "date",
    });

    expect(toSchema(inferSchema("09:24:15"))).toStrictEqual({
      type: "string",
      format: "time",
    });

    expect(toSchema(inferSchema("192.168.1.0", inferSchema("2019-01-01")))).toStrictEqual({
      type: "string",
    });

    expect(toSchema(inferSchema("2020-12-01", inferSchema("2019-01-01")))).toStrictEqual({
      type: "string",
      format: "date",
    });
  });

  it("should not infer string formats when all strings don't have the same format", () => {
    expect(toSchema(inferSchema("2020-12-01", inferSchema("this is not formatted")))).toStrictEqual(
      { type: "string" },
    );

    expect(toSchema(inferSchema("this is not formatted", inferSchema("2020-12-01")))).toStrictEqual(
      { type: "string" },
    );

    expect(toSchema(inferSchema("US", inferSchema("2020-12-01")))).toStrictEqual({
      type: "string",
    });

    expect(toSchema(inferSchema("google.com", inferSchema("2020-12-01")))).toStrictEqual({
      type: "string",
    });

    expect(
      toSchema(inferSchema("ee487ff7-374d-4e50-9b72-ac51e4459c5f", inferSchema("2020-12-01"))),
    ).toStrictEqual({
      type: "string",
    });

    expect(toSchema(inferSchema("http://google.com", inferSchema("2020-12-01")))).toStrictEqual({
      type: "string",
    });

    expect(toSchema(inferSchema("+447456001234", inferSchema("2020-12-01")))).toStrictEqual({
      type: "string",
    });

    expect(toSchema(inferSchema("ES", inferSchema("2020-12-01")))).toStrictEqual({
      type: "string",
    });
  });

  it("Should infer anyOf when a string is followed by something else", () => {
    expect(toSchema(inferSchema(1, inferSchema("hello world")))).toStrictEqual({
      anyOf: [{ type: "string" }, { type: "integer" }],
    });
  });
});

describe("arrays", () => {
  test("inferring a homogeneous array", () => {
    expect(toSchema(inferSchema([1, 2, 3]))).toStrictEqual({
      type: "array",
      items: { type: "integer" },
    });
  });

  test("inferring a heterogeneous array", () => {
    expect(toSchema(inferSchema([1, "hello world", false]))).toStrictEqual({
      type: "array",
      items: { anyOf: [{ type: "integer" }, { type: "string" }, { type: "boolean" }] },
    });
  });

  test("inferring an array multiple times", () => {
    expect(toSchema(inferSchema([4, 5, 6], inferSchema([1, 2, 3])))).toStrictEqual({
      type: "array",
      items: { type: "integer" },
    });
  });

  test("following an array inferring with something other than an array", () => {
    expect(toSchema(inferSchema("hello world", inferSchema([1, 2, 3])))).toStrictEqual({
      anyOf: [
        {
          type: "array",
          items: { type: "integer" },
        },
        { type: "string" },
      ],
    });
  });
});

describe("objects", () => {
  test("inferring an object", () => {
    expect(toSchema(inferSchema({ foo: "bar" }))).toStrictEqual({
      type: "object",
      properties: {
        foo: { type: "string" },
      },
      required: ["foo"],
    });
  });

  test("inferring optional properties", () => {
    let schema = inferSchema({ foo: "bar", baz: "qux" });
    schema = inferSchema({ foo: "bar", banana: 1 }, schema);

    expect(toSchema(schema)).toStrictEqual({
      type: "object",
      properties: {
        foo: { type: "string" },
        baz: { type: "string" },
        banana: { type: "integer" },
      },
      required: ["foo"],
    });

    schema = inferSchema({ baz: "qux" }, schema);

    expect(toSchema(schema)).toStrictEqual({
      type: "object",
      properties: {
        foo: { type: "string" },
        baz: { type: "string" },
        banana: { type: "integer" },
      },
    });
  });

  test("following an object inferring with something other than an object", () => {
    expect(toSchema(inferSchema("hello world", inferSchema({})))).toStrictEqual({
      anyOf: [{ type: "object" }, { type: "string" }],
    });
  });
});

describe("array of objects", () => {
  test("infers the items of the array as an object", () => {
    expect(toSchema(inferSchema([{ foo: "bar", baz: "qux" }, { foo: "bar" }]))).toStrictEqual({
      type: "array",
      items: {
        type: "object",
        properties: {
          foo: { type: "string" },
          baz: { type: "string" },
        },
        required: ["foo"],
      },
    });
  });
});

describe("nullables", () => {
  test("infers the schema as allowing null as well as another type", () => {
    const schema = inferSchema([1, 2, null, 3]);

    expect(toSchema(schema)).toStrictEqual({
      type: "array",
      items: {
        type: ["integer", "null"],
      },
    });
  });

  test("infers just a single null type if multiple nulls are found", () => {
    const schema = inferSchema([1, 2, null, 3, null, 4]);

    expect(toSchema(schema)).toStrictEqual({
      type: "array",
      items: {
        type: ["integer", "null"],
      },
    });
  });
});

describe("real world tests", () => {
  test("infers the airtable schema correctly", () => {
    expect(fixtureToSchema("airtable")).toMatchSnapshot();
  });

  test("infers the mux-list-delivery-usage schema correctly", () => {
    expect(fixtureToSchema("mux-list-delivery-usage")).toMatchSnapshot();
  });

  test("infers the tweets schema correctly", () => {
    expect(fixtureToSchema("tweets")).toMatchSnapshot();
  });
});

import Ajv2020 from "ajv/dist/2020";
import addFormats from "ajv-formats";
const ajv = new Ajv2020();
addFormats(ajv);

describe("validation", () => {
  it("should pass validation if given the same json the schema was inferred from", () => {
    const schema = fixtureToSchema("tweets");
    const validate = ajv.compile(schema);
    const tweetsJson = readFixture("tweets");

    expect(validate(tweetsJson)).toBe(true);
  });
});
