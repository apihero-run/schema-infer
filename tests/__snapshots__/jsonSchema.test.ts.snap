// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`real world tests infers the airtable schema correctly 1`] = `
Object {
  "properties": Object {
    "offset": Object {
      "type": "string",
    },
    "records": Object {
      "items": Object {
        "properties": Object {
          "createdTime": Object {
            "format": "date-time",
            "type": "string",
          },
          "fields": Object {
            "properties": Object {
              "Description": Object {
                "type": "string",
              },
              "Face Images": Object {
                "items": Object {
                  "properties": Object {
                    "filename": Object {
                      "type": "string",
                    },
                    "height": Object {
                      "type": "integer",
                    },
                    "id": Object {
                      "type": "string",
                    },
                    "size": Object {
                      "type": "integer",
                    },
                    "thumbnails": Object {
                      "properties": Object {
                        "full": Object {
                          "properties": Object {
                            "height": Object {
                              "type": "integer",
                            },
                            "url": Object {
                              "format": "uri",
                              "type": "string",
                            },
                            "width": Object {
                              "type": "integer",
                            },
                          },
                          "required": Array [
                            "url",
                            "width",
                            "height",
                          ],
                          "type": "object",
                        },
                        "large": Object {
                          "properties": Object {
                            "height": Object {
                              "type": "integer",
                            },
                            "url": Object {
                              "format": "uri",
                              "type": "string",
                            },
                            "width": Object {
                              "type": "integer",
                            },
                          },
                          "required": Array [
                            "url",
                            "width",
                            "height",
                          ],
                          "type": "object",
                        },
                        "small": Object {
                          "properties": Object {
                            "height": Object {
                              "type": "integer",
                            },
                            "url": Object {
                              "format": "uri",
                              "type": "string",
                            },
                            "width": Object {
                              "type": "integer",
                            },
                          },
                          "required": Array [
                            "url",
                            "width",
                            "height",
                          ],
                          "type": "object",
                        },
                      },
                      "required": Array [
                        "small",
                        "large",
                        "full",
                      ],
                      "type": "object",
                    },
                    "type": Object {
                      "type": "string",
                    },
                    "url": Object {
                      "format": "uri",
                      "type": "string",
                    },
                    "width": Object {
                      "type": "integer",
                    },
                  },
                  "required": Array [
                    "id",
                    "width",
                    "height",
                    "url",
                    "filename",
                    "size",
                    "type",
                    "thumbnails",
                  ],
                  "type": "object",
                },
                "type": "array",
              },
              "Faces": Object {
                "items": Object {
                  "type": "string",
                },
                "type": "array",
              },
              "ID": Object {
                "type": "integer",
              },
              "Image": Object {
                "items": Object {
                  "properties": Object {
                    "filename": Object {
                      "type": "string",
                    },
                    "height": Object {
                      "type": "integer",
                    },
                    "id": Object {
                      "type": "string",
                    },
                    "size": Object {
                      "type": "integer",
                    },
                    "thumbnails": Object {
                      "properties": Object {
                        "full": Object {
                          "properties": Object {
                            "height": Object {
                              "type": "integer",
                            },
                            "url": Object {
                              "format": "uri",
                              "type": "string",
                            },
                            "width": Object {
                              "type": "integer",
                            },
                          },
                          "required": Array [
                            "url",
                            "width",
                            "height",
                          ],
                          "type": "object",
                        },
                        "large": Object {
                          "properties": Object {
                            "height": Object {
                              "type": "integer",
                            },
                            "url": Object {
                              "format": "uri",
                              "type": "string",
                            },
                            "width": Object {
                              "type": "integer",
                            },
                          },
                          "required": Array [
                            "url",
                            "width",
                            "height",
                          ],
                          "type": "object",
                        },
                        "small": Object {
                          "properties": Object {
                            "height": Object {
                              "type": "integer",
                            },
                            "url": Object {
                              "format": "uri",
                              "type": "string",
                            },
                            "width": Object {
                              "type": "integer",
                            },
                          },
                          "required": Array [
                            "url",
                            "width",
                            "height",
                          ],
                          "type": "object",
                        },
                      },
                      "required": Array [
                        "small",
                        "large",
                        "full",
                      ],
                      "type": "object",
                    },
                    "type": Object {
                      "type": "string",
                    },
                    "url": Object {
                      "format": "uri",
                      "type": "string",
                    },
                    "width": Object {
                      "type": "integer",
                    },
                  },
                  "required": Array [
                    "id",
                    "width",
                    "height",
                    "url",
                    "filename",
                    "size",
                    "type",
                    "thumbnails",
                  ],
                  "type": "object",
                },
                "type": "array",
              },
              "Notes": Object {
                "type": "string",
              },
              "People": Object {
                "items": Object {
                  "type": "string",
                },
                "type": "array",
              },
              "Status": Object {
                "type": "string",
              },
              "Tags": Object {
                "items": Object {
                  "type": "string",
                },
                "type": "array",
              },
            },
            "required": Array [
              "Tags",
              "ID",
              "Description",
              "Faces",
              "Image",
              "Status",
              "Notes",
              "People",
              "Face Images",
            ],
            "type": "object",
          },
          "id": Object {
            "type": "string",
          },
        },
        "required": Array [
          "id",
          "fields",
          "createdTime",
        ],
        "type": "object",
      },
      "type": "array",
    },
  },
  "required": Array [
    "records",
    "offset",
  ],
  "type": "object",
}
`;

exports[`real world tests infers the mux-list-delivery-usage schema correctly 1`] = `
Object {
  "properties": Object {
    "data": Object {
      "items": Object {
        "properties": Object {
          "asset_duration": Object {
            "type": "number",
          },
          "asset_id": Object {
            "type": "string",
          },
          "asset_state": Object {
            "type": "string",
          },
          "created_at": Object {
            "type": "string",
          },
          "deleted_at": Object {
            "type": "string",
          },
          "delivered_seconds": Object {
            "type": "number",
          },
          "live_stream_id": Object {
            "type": "string",
          },
        },
        "required": Array [
          "delivered_seconds",
          "deleted_at",
          "created_at",
          "asset_state",
          "asset_id",
          "asset_duration",
        ],
        "type": "object",
      },
      "type": "array",
    },
    "limit": Object {
      "type": "integer",
    },
    "page": Object {
      "type": "integer",
    },
    "timeframe": Object {
      "items": Object {
        "type": "integer",
      },
      "type": "array",
    },
    "total_row_count": Object {
      "type": "integer",
    },
  },
  "required": Array [
    "total_row_count",
    "timeframe",
    "page",
    "limit",
    "data",
  ],
  "type": "object",
}
`;

exports[`real world tests infers the tweets schema correctly 1`] = `
Object {
  "items": Object {
    "properties": Object {
      "data": Object {
        "items": Object {
          "properties": Object {
            "author_id": Object {
              "type": "string",
            },
            "context_annotations": Object {
              "items": Object {
                "properties": Object {
                  "domain": Object {
                    "properties": Object {
                      "description": Object {
                        "type": "string",
                      },
                      "id": Object {
                        "type": "string",
                      },
                      "name": Object {
                        "type": "string",
                      },
                    },
                    "required": Array [
                      "id",
                      "name",
                      "description",
                    ],
                    "type": "object",
                  },
                  "entity": Object {
                    "properties": Object {
                      "id": Object {
                        "type": "string",
                      },
                      "name": Object {
                        "type": "string",
                      },
                    },
                    "required": Array [
                      "id",
                      "name",
                    ],
                    "type": "object",
                  },
                },
                "required": Array [
                  "domain",
                  "entity",
                ],
                "type": "object",
              },
              "type": "array",
            },
            "conversation_id": Object {
              "type": "string",
            },
            "created_at": Object {
              "format": "date-time",
              "type": "string",
            },
            "entities": Object {
              "properties": Object {
                "annotations": Object {
                  "items": Object {
                    "properties": Object {
                      "end": Object {
                        "type": "integer",
                      },
                      "normalized_text": Object {
                        "type": "string",
                      },
                      "probability": Object {
                        "type": "number",
                      },
                      "start": Object {
                        "type": "integer",
                      },
                      "type": Object {
                        "type": "string",
                      },
                    },
                    "required": Array [
                      "start",
                      "end",
                      "probability",
                      "type",
                      "normalized_text",
                    ],
                    "type": "object",
                  },
                  "type": "array",
                },
                "mentions": Object {
                  "items": Object {
                    "properties": Object {
                      "end": Object {
                        "type": "integer",
                      },
                      "start": Object {
                        "type": "integer",
                      },
                      "username": Object {
                        "type": "string",
                      },
                    },
                    "required": Array [
                      "start",
                      "end",
                      "username",
                    ],
                    "type": "object",
                  },
                  "type": "array",
                },
                "urls": Object {
                  "items": Object {
                    "properties": Object {
                      "description": Object {
                        "type": "string",
                      },
                      "display_url": Object {
                        "type": "string",
                      },
                      "end": Object {
                        "type": "integer",
                      },
                      "expanded_url": Object {
                        "format": "uri",
                        "type": "string",
                      },
                      "images": Object {
                        "items": Object {
                          "properties": Object {
                            "height": Object {
                              "type": "integer",
                            },
                            "url": Object {
                              "format": "uri",
                              "type": "string",
                            },
                            "width": Object {
                              "type": "integer",
                            },
                          },
                          "required": Array [
                            "url",
                            "width",
                            "height",
                          ],
                          "type": "object",
                        },
                        "type": "array",
                      },
                      "start": Object {
                        "type": "integer",
                      },
                      "status": Object {
                        "type": "integer",
                      },
                      "title": Object {
                        "type": "string",
                      },
                      "unwound_url": Object {
                        "format": "uri",
                        "type": "string",
                      },
                      "url": Object {
                        "format": "uri",
                        "type": "string",
                      },
                    },
                    "required": Array [
                      "start",
                      "end",
                      "url",
                      "expanded_url",
                      "display_url",
                      "status",
                      "title",
                      "description",
                      "unwound_url",
                    ],
                    "type": "object",
                  },
                  "type": "array",
                },
              },
              "required": Array [
                "urls",
              ],
              "type": "object",
            },
            "id": Object {
              "type": "string",
            },
            "in_reply_to_user_id": Object {
              "type": "string",
            },
            "lang": Object {
              "type": "string",
            },
            "possibly_sensitive": Object {
              "type": "boolean",
            },
            "public_metrics": Object {
              "properties": Object {
                "like_count": Object {
                  "type": "integer",
                },
                "quote_count": Object {
                  "type": "integer",
                },
                "reply_count": Object {
                  "type": "integer",
                },
                "retweet_count": Object {
                  "type": "integer",
                },
              },
              "required": Array [
                "retweet_count",
                "reply_count",
                "like_count",
                "quote_count",
              ],
              "type": "object",
            },
            "referenced_tweets": Object {
              "items": Object {
                "properties": Object {
                  "id": Object {
                    "type": "string",
                  },
                  "type": Object {
                    "type": "string",
                  },
                },
                "required": Array [
                  "type",
                  "id",
                ],
                "type": "object",
              },
              "type": "array",
            },
            "source": Object {
              "type": "string",
            },
            "text": Object {
              "type": "string",
            },
          },
          "required": Array [
            "conversation_id",
            "id",
            "possibly_sensitive",
            "public_metrics",
            "entities",
            "text",
            "in_reply_to_user_id",
            "created_at",
            "author_id",
            "referenced_tweets",
            "lang",
            "source",
          ],
          "type": "object",
        },
        "type": "array",
      },
      "includes": Object {
        "properties": Object {
          "tweets": Object {
            "items": Object {
              "properties": Object {
                "attachments": Object {
                  "properties": Object {
                    "media_keys": Object {
                      "items": Object {
                        "type": "string",
                      },
                      "type": "array",
                    },
                  },
                  "required": Array [
                    "media_keys",
                  ],
                  "type": "object",
                },
                "author_id": Object {
                  "type": "string",
                },
                "conversation_id": Object {
                  "type": "string",
                },
                "created_at": Object {
                  "format": "date-time",
                  "type": "string",
                },
                "entities": Object {
                  "properties": Object {
                    "annotations": Object {
                      "items": Object {
                        "properties": Object {
                          "end": Object {
                            "type": "integer",
                          },
                          "normalized_text": Object {
                            "type": "string",
                          },
                          "probability": Object {
                            "type": "number",
                          },
                          "start": Object {
                            "type": "integer",
                          },
                          "type": Object {
                            "type": "string",
                          },
                        },
                        "required": Array [
                          "start",
                          "end",
                          "probability",
                          "type",
                          "normalized_text",
                        ],
                        "type": "object",
                      },
                      "type": "array",
                    },
                    "hashtags": Object {
                      "items": Object {
                        "properties": Object {
                          "end": Object {
                            "type": "integer",
                          },
                          "start": Object {
                            "type": "integer",
                          },
                          "tag": Object {
                            "type": "string",
                          },
                        },
                        "required": Array [
                          "start",
                          "end",
                          "tag",
                        ],
                        "type": "object",
                      },
                      "type": "array",
                    },
                    "mentions": Object {
                      "items": Object {
                        "properties": Object {
                          "end": Object {
                            "type": "integer",
                          },
                          "start": Object {
                            "type": "integer",
                          },
                          "username": Object {
                            "type": "string",
                          },
                        },
                        "required": Array [
                          "start",
                          "end",
                          "username",
                        ],
                        "type": "object",
                      },
                      "type": "array",
                    },
                    "urls": Object {
                      "items": Object {
                        "properties": Object {
                          "display_url": Object {
                            "type": "string",
                          },
                          "end": Object {
                            "type": "integer",
                          },
                          "expanded_url": Object {
                            "format": "uri",
                            "type": "string",
                          },
                          "start": Object {
                            "type": "integer",
                          },
                          "url": Object {
                            "format": "uri",
                            "type": "string",
                          },
                        },
                        "required": Array [
                          "start",
                          "end",
                          "url",
                          "expanded_url",
                          "display_url",
                        ],
                        "type": "object",
                      },
                      "type": "array",
                    },
                  },
                  "required": Array [
                    "mentions",
                  ],
                  "type": "object",
                },
                "id": Object {
                  "type": "string",
                },
                "lang": Object {
                  "type": "string",
                },
                "possibly_sensitive": Object {
                  "type": "boolean",
                },
                "public_metrics": Object {
                  "properties": Object {
                    "like_count": Object {
                      "type": "integer",
                    },
                    "quote_count": Object {
                      "type": "integer",
                    },
                    "reply_count": Object {
                      "type": "integer",
                    },
                    "retweet_count": Object {
                      "type": "integer",
                    },
                  },
                  "required": Array [
                    "retweet_count",
                    "reply_count",
                    "like_count",
                    "quote_count",
                  ],
                  "type": "object",
                },
                "source": Object {
                  "type": "string",
                },
                "text": Object {
                  "type": "string",
                },
              },
              "required": Array [
                "conversation_id",
                "id",
                "possibly_sensitive",
                "public_metrics",
                "entities",
                "text",
                "created_at",
                "author_id",
                "lang",
                "source",
              ],
              "type": "object",
            },
            "type": "array",
          },
          "users": Object {
            "items": Object {
              "properties": Object {
                "created_at": Object {
                  "format": "date-time",
                  "type": "string",
                },
                "description": Object {
                  "type": "string",
                },
                "entities": Object {
                  "properties": Object {
                    "description": Object {
                      "properties": Object {
                        "hashtags": Object {
                          "items": Object {
                            "properties": Object {
                              "end": Object {
                                "type": "integer",
                              },
                              "start": Object {
                                "type": "integer",
                              },
                              "tag": Object {
                                "type": "string",
                              },
                            },
                            "required": Array [
                              "start",
                              "end",
                              "tag",
                            ],
                            "type": "object",
                          },
                          "type": "array",
                        },
                        "mentions": Object {
                          "items": Object {
                            "properties": Object {
                              "end": Object {
                                "type": "integer",
                              },
                              "start": Object {
                                "type": "integer",
                              },
                              "username": Object {
                                "type": "string",
                              },
                            },
                            "required": Array [
                              "start",
                              "end",
                              "username",
                            ],
                            "type": "object",
                          },
                          "type": "array",
                        },
                      },
                      "type": "object",
                    },
                    "url": Object {
                      "properties": Object {
                        "urls": Object {
                          "items": Object {
                            "properties": Object {
                              "display_url": Object {
                                "type": "string",
                              },
                              "end": Object {
                                "type": "integer",
                              },
                              "expanded_url": Object {
                                "format": "uri",
                                "type": "string",
                              },
                              "start": Object {
                                "type": "integer",
                              },
                              "url": Object {
                                "format": "uri",
                                "type": "string",
                              },
                            },
                            "required": Array [
                              "start",
                              "end",
                              "url",
                              "expanded_url",
                              "display_url",
                            ],
                            "type": "object",
                          },
                          "type": "array",
                        },
                      },
                      "required": Array [
                        "urls",
                      ],
                      "type": "object",
                    },
                  },
                  "required": Array [
                    "url",
                    "description",
                  ],
                  "type": "object",
                },
                "id": Object {
                  "type": "string",
                },
                "location": Object {
                  "type": "string",
                },
                "name": Object {
                  "type": "string",
                },
                "pinned_tweet_id": Object {
                  "type": "string",
                },
                "profile_image_url": Object {
                  "format": "uri",
                  "type": "string",
                },
                "protected": Object {
                  "type": "boolean",
                },
                "public_metrics": Object {
                  "properties": Object {
                    "followers_count": Object {
                      "type": "integer",
                    },
                    "following_count": Object {
                      "type": "integer",
                    },
                    "listed_count": Object {
                      "type": "integer",
                    },
                    "tweet_count": Object {
                      "type": "integer",
                    },
                  },
                  "required": Array [
                    "followers_count",
                    "following_count",
                    "tweet_count",
                    "listed_count",
                  ],
                  "type": "object",
                },
                "url": Object {
                  "format": "uri",
                  "type": "string",
                },
                "username": Object {
                  "type": "string",
                },
                "verified": Object {
                  "type": "boolean",
                },
              },
              "required": Array [
                "created_at",
                "profile_image_url",
                "entities",
                "id",
                "verified",
                "location",
                "description",
                "username",
                "public_metrics",
                "name",
                "url",
                "protected",
              ],
              "type": "object",
            },
            "type": "array",
          },
        },
        "required": Array [
          "users",
          "tweets",
        ],
        "type": "object",
      },
    },
    "required": Array [
      "data",
      "includes",
    ],
    "type": "object",
  },
  "type": "array",
}
`;
